WIP: added travis CI testing for prplmesh

slyusar003 opened the pull request:
- added test param in maptools.py
- added check param in prplmesh-utils.py

Signed-off-by: Ostap <slyusar.ostap@gmail.com>

slyusar003 committed:
WIP: added travis CI testing for prplmesh

- added test param in maptools.py
- added check param in prplmesh-utils.py

Signed-off-by: Ostap <slyusar.ostap@gmail.com>

Commit:
tools: docker - add 'runner' docker support

Add a "runner" docker support for running prplMesh in docker containers.
Using this, we can now run 2 containers, one with a controller, the
other with an agent.
The 2 containers are connected using a docker network (bridge), so can
communicate + we can now sniff the bridge from the host to see 1905
packets.

First, build prplMesh which will be used by the containers
(build/install directory will be the only directory mapped to the
containers).

Next, build the docker image using `./image-build.sh`.

Then, create a docker network (bridge) `sudo docker network create
prplMesh-net`. Write down the bridge IP address (using `ip a` or
`sudo docker network inspect prplMesh-net`) - we will need it for the next step.

Finally, run 2 containers, each with a different IP address in the docker
network subnet. So for example, if the subnet is 172.20.0.0/16 and the
docker bridge in the host is 172.20.0.1:
Run Agent+Controller container - `./run.sh 172.20.0.2`
Run Agent container - `./run.sh 172.20.0.3`

Now you will get into each container shell and can start prplMesh:
Running the Agent+controller:
```
sudo ./scripts/prplmesh_utils.sh start -s
tail -F /tmp/beerocks/logs/beerocks_controller.log
```

Running the agent:
```
sudo ./scripts/prplmesh_utils.sh start -s -m A
tail -F /tmp/beerocks/logs/beerocks_agent.log
```

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on tools/maptools.py:
```diff
@@ -69,13 +72,17 @@ def __deploy__(self):
             'deploy': self.__deploy__,
-            'copy': self.__copy__
+            'copy': self.__copy__,
+            'test': self.__test__,
```
Great idea to add a `test` command to maptools.

Note that this should be in a separate commit from the changes to prplmesh_utils.sh and the changes to .travis.yml. So in the end, there should be 3 commits I think:

1. Add `check` to prplmesh_utils.sh
2. Add `test` to maptools
3. Add call to `test` to travis.yml

arnout commented on tools/maptools.py:
```diff
@@ -69,13 +72,17 @@ def __deploy__(self):
     def __copy__(self):
         mapcopy(self.args)
 
+    def __test__(self):
```
Hm, it's wrong to make this a dunder method, since dunder methods are supposed to be reserved for the language. Still, the rest of the class does it like that, so for consistency it's good to do the same.

arnout commented on tools/commands/test.py:
```diff
@@ -0,0 +1,49 @@
+logger = logging.getLogger("test")
+test_targets=['start', 'stop', 'restart', 'status', 'check']
+
+class prpltester(object):
```
I don't think it's useful to make this a separate class, just merge it into the maptest class.

slyusar003 replied:
okay, I was thinking that there will be more tests in the future but I will remove a class for now.

arnout commented on tools/commands/test.py:
```diff
@@ -0,0 +1,49 @@
+        if 'check' in commands:
+            tester.run_cmd('restart')
+            tester.run_cmd('check')
+            tester.run_cmd('stop')
```
Originally I thought, better just do `for c in commands: tester.run_cmd(c)`. However, this really makes a lot of sense. :+1:

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -104,8 +104,13 @@ status_function() {
 
+check_function() {
+    echo "$0: check"
+    cat "/tmp/$USER/beerocks/logs/beerocks_agent.log" |grep 'CONNECTED --> OPERATIONAL'
```
You don't need `cat` here, just
```suggestion
    if ! grep -q 'CONNECTED --> OPERATIONAL' "/tmp/$USER/beerocks/logs/beerocks_agent.log"; then
        err "check failed"
        exit 1
    fi
```

Commit:
scripts: prplmesh_utils.sh docker adaptations

Add an option for not running platform init / deinit since in the docker
runner we are setting the bridge and wlan interfaces in advance.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
framework: update tmp path

When running multiple prplMesh instances on the same linux host by
multiple users (normal development flow using a shared server), the
framework build system uses /tmp/$USER as the tmp path prefix used for
logs and the local bus endpoints.

In containers, this is not needed, so convert this into an option -
"SHARED_SERVER", default ON (for linux platform), and when building for
containers the user can set this to OFF.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
controller: update tmp path

When running multiple prplMesh instances on the same linux host by
multiple users (normal development flow using a shared server), the
controller build system uses /tmp/$USER as the tmp path prefix used for
logs.

In containers, this is not needed, so convert this into an option -
"SHARED_SERVER", default ON (for linux platform), and when building for
containers the user can set this to OFF.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: update tmp path

When running multiple prplMesh instances on the same linux host by
multiple users (normal development flow using a shared server), the
agent's build system uses /tmp/$USER as the tmp path prefix used for
logs.

In containers, this is not needed, so convert this into an option -
"SHARED_SERVER", default ON (for linux platform), and when building for
containers the user can set this to OFF.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: docker - add 'runner' docker README.md

Add README.md for runner docker.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

slyusar003 committed:
fixup! WIP: added travis CI testing for prplmesh

Signed-off-by: Ostap <slyusar.ostap@gmail.com>

tomereli commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -104,8 +104,16 @@ status_function() {
     ps -aux | grep local_bus
 }
 
+check_function() {
```
I think it is better to add this into the status function

tomereli replied:
OPERATIONAL in the main agent doesn't mean the radio agents are operational - the status check should first indeed check that the main agent is operational, and then grep beerocks_agent_wlan0.log to see it got to a working state - for now, in DUMMY mode, we know the son slave is functional if we see `ACTION_AP_MANAGER_HOSTAP_CSA_NOTIFICATION`

tomereli commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -104,8 +104,16 @@ status_function() {
+
 usage() {
-    echo "usage: $(basename $0) {start|stop|restart|status} [-hv]"
+    echo "usage: $(basename $0) {start|stop|restart|status|check} [-hv]"
```
no need for `check` - should be part of status, return code defines if system is operational, and we can add a --quiet flag if we want to get only the check status

arnout replied:
Requires more changes to status_function, however, because currently it doesn't set the exit code, nor does it use BTL_TYPE or PRPLMESH_MODE to determine which daemons should be running.

tomereli commented on tools/commands/test.py:
```diff
@@ -0,0 +1,49 @@
+    def configure_parser(parser=argparse.ArgumentParser(prog='test')):
+        parser.help = "multiap_sw standalone test module"
+        parser.add_argument("--verbose", "-v", action="store_true", help="verbosity on")
+        parser.add_argument('-c', '--commands', choices=test_targets, nargs='+', default=['check'], help="test command (default is check)")
```
I think a better name if you take my comments from above is to call this 'status'
`./maptools.py test -c status`

Then later we can add more tests, i.e. `./maptools.py test -c ctest`, etc.

slyusar003 committed:
fixup! WIP: added travis CI testing for prplmesh

Signed-off-by: Ostap <slyusar.ostap@gmail.com>
